### ISSUE TITLE
[question] What's the intended way to work with makefiles without autotools in 2.0

### ISSUE THREAD

**Issue by daltonv** [conan-io/conan#13394]
### What is your question?

I'm new to Conan with 2.0 and my team is currently using custom makefiles without autotools to build our projects. Googling around I end up on 1.x docs of Conan where there does seem to be built in generators for makefiles. So how do I work with makefiles in 2.0?

Packaging at least seems easy. I can just not have a `build()` function and use a wrapper script to build with make first and then call `conan create`. Maybe I could have the `build()` function call the python `subprocess` library to handle the build itself though?

However consuming conan packages is not as straight forward. Should I just be using the `VirtualBuildEnv` and relying on env vars in my makefiles?

Are there any good example projects that build with makefiles and conan 2.0?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @daltonv 

Not sure if you are aware of the ``conan new autotools_lib -d name=pkg -d version=0.1``.
It is true that it is an autotools project, not a pure Makefile one, but I think stripping out the autotools part should be relatively easy and the overall approach should be a good start.

> I can just not have a build() function and use a wrapper script to build with make first and then call conan create. 

Sorry, I don't know what you mean. Typically running ``self.run("make ...")`` should be the way to go. In general running ``subprocess`` yourself is not recommended, but using the ``self.run()``

Yes, the ``test_package/conanfile.py`` inside the template would be almost a typical consumer project. Not ``VirtualBuildEnv``, but depending on what you are doing ``AutotoolsDeps`` or ``PkgConfigDeps`` should probably be the way to go. Even if it is called ``AutotoolsDeps`` it is mostly generating environment scripts with the environment variables necessary to locate the dependencies, nothing really strictly of "autotools" only. Probably same for ``AutotoolsToolchain``.


I'd recommend giving a try to the template, removing the autotools bits, and if you have any issues, let us know and we could try to put together an example. Thanks!

**Comment by memsharded** [comment]
Actually, this might be useful for some other users, maybe we want to make a ``conan new makefile_lib`` (and _exe) templates too.

**Comment by daltonv** [comment]
Ah I didn't see the `ConanFile` class has a `self.run()` method. That makes it easy.

I'll try out the autotools project and report back.

I definitely think that a makefile example would be helpful. I'm working on embedded projects, and while I can control my build system, many embedded sdks and libraries still use plain makefiles to build.

**Comment by memsharded** [comment]
> Ah I didn't see the ConanFile class has a self.run() method. That makes it easy.

Yes, and beyond easier, the ``self.run()`` might do other stuff, like automatically inject environment variables scripts.

> I'll try out the autotools project and report back.

I definitely think that a makefile example would be helpful. I'm working on embedded projects, and while I can control my build system, many embedded sdks and libraries still use plain makefiles to build.

Perfect, looking forward your feedback, please share you findings and we will propose a new pure Makefiles template. Thanks!

**Comment by daltonv** [comment]
Okay it was actually pretty easy to get the makefiles running now that I know where to start! I used the `AutotoolsDeps` generator as suggested.

A couple points of feedback:

## Docs

Initially I thought I just missed something in the docs about the `self.run()` method, and I did, but I feel like it could have been more obvious. It's indirectly talked about in the [build() tutorial](https://docs.conan.io/2/tutorial/creating_packages/build_packages.html#changes-introduced-in-the-recipe) but I missed that on my first glance. The only other obvious place I saw it was in reference material for [build()](https://docs.conan.io/2/reference/conanfile/methods/build.html). I think the docs should at least have a reference section about the `run()` command as currently you have to look at the source code to understand the args. Additionally a makfile example would probably make it more obvious that the `run()` command exists.

## [tool_requires] flags

Currently the `AutotoolsDeps` generator does not include dependencies from [tool_requires] settings. The [docs](https://docs.conan.io/2/reference/tools/gnu/autotoolsdeps.html#generated-files) seem to imply that this will be included in the future. I really hope so, because I want to package the arm gcc compiler to be used in all my projects, and there are some arm specific cflags I want all projects to use. So it would be nice if the compiler package provides those to downstream projects.

I can work around this by just adding those settings directly in the makefile, but it's much cleaner to cross compiler for x86 and various arm archs in `CMake` land as the [tool_requires] `cpp_info` settings are carried over. 

Even if I put the cflags in my profile file `AutotoolsDeps` doesn't pick them up. Example:
```ini
[conf]
tools.build:cflags+=["-gstrict-dwarf", "-march=armv6-m"]
```

Overall though it's pretty easy to setup makefile projects with `AutotoolsDeps` + `self.run(make)`. I think an example that shows this would be very nice. 

Just need to generate with `AutotoolsDeps` and call `self.run(make)` the `package()` function will likely be custom for each makefile project.

**Comment by gabeblack** [comment]
Within the files created by `conan new autotools_lib` there is a folder called `test_package` which I believe is an example of a project that "consumes" the library built. However, it is still autotools... I mean `autotools` is required if I want the consumer test_package to build without running `conan build ...` right?

The same is true with other generators (e.g., cmake), `conan install .` on the consumer project will create the files that then my consumer's `CMakeLists.txt` can include them in order to pull in the library helper vars needed to provide the CPPFLAGS/LDFLAGS/etc for building. With those generators, I can then run `cmake` followed by  `make` directly to build my consumer project (i.e., I don't have to run `conan build`).

I don't see how the same thing can be done with the `autotools_lib` template and a standard Makefile project. I don't see any files that are generated by `Autotoolsdeps` generator that a `Makefile` could include to build the consumer project where I simply run "make" and the consumer project builds.  Am I missing something? Or maybe I'm misunderstanding the workflow?

**Comment by daltonv** [comment]
Hey @gabeblack. The generated `conanbuild.sh` script and friends from `AutotoolsDeps` can inject info into the makefile. This is because makefiles will consume environment variables.

So the correct flow will be:
1. Run `conan install .`
2. Run `source conanbuild.sh` - This loads the shell scripts that set env vars for stuff like `CC`, `CFLAGS`, etc.
3. Run `make` - Your makefile should be using the special env vars. Like `$(CC) $(CPPFLAGS) $(CFLAGS) -c $< -o $@`.

If you add `self. run(make)` to your `conanfile.py` (and correctly generated the `AutotoolsDeps`) the `run()` method automagically loads the env vars for you, so you skip step 2, and `conan build .` handles it for you.

**Comment by JamesMTSloan** [comment]
I've had a very similar experience to the OP and landing on this issue was the help I needed since I was getting the impression that since it was in the 1.x docs but not the 2.x that this kind of support had been removed.

I started from the Autotools template and removed things from there, although I still use `Autotools.make()` and `Autotools.install()` instead of `self.run()` as all they do is call the correct things anyway. Having mention of Autotools at all does make the contents of my `conanfile.py` a bit misleading, but I am also using `generators = "AutotoolsToolchain", "AutotoolsDeps"` to get the environment set up. Maybe some sort of alias or explicit support would be good.

Anyway, the upshot is that I can retro-fit old `make` based projects to use Conan 2.0 successfully with this approach. Just posting this to echo the OP's thoughts, emphasis the use-case, and reassure anyone else coming across this issue.

**Comment by iskunk** [comment]
Hello folks,

I am currently looking into this issue myself. One problem with the `AutotoolsDeps` generator (and many others) is that it only aggregates the variables of multiple packages. For example, in the generator output, `CPPFLAGS` contains all the `-Idirectory` flags for the various packages, concatenated together. Which, to be sure, is fine for many use cases. But if you need the `CPPFLAGS` variable specifically for the `foobaz` package, then you're out of luck---there is no `CONAN_INCLUDE_DIRS_FOOBAZ` variable that can be referenced. You'll have to do a `conan install` for that package alone.

Also, it's not clear to me if/how the `AutotoolsDeps` generator addresses some of the other variables provided by the `make` one, like `CONAN_RES_DIRS_FOOBAZ`. As far as I can tell, it only deals with the official GNU-recognized environment variables.

Is there any work in progress on a Make-compatible generator that will provide per-package variables? Would a PR to port the old `make` generator to 2.0 (or perhaps develop a new one) be OK in principle?

**Comment by gabeblack** [comment]
@iskunk -- I actually ran in to the same thing. So yes, you can make the `Autotools`/`AutotoolsDeps` work for many makefile packages, but it is incomplete for the reason you gave. 

I ended up having to sort of write my own generator to do what you and I were needing. The following will create a `CONAN_<pkg_name>_PKG_FOLDER` which from that you can tag on "include", "libs", etc to get the include, libs, etc folders. Or you can extend this implementation to make the package-specific variables. I also had it rename the global CPPFLAGS to be CONAN_CPPFLAGS, LDFLAGS to CONAN_LDFLAGS, etc, etc. In my use-case, I didn't want the conan CPPFLAGS to be used for my entire project. Just parts. So I renamed the global variables conan creates and prefixed them with 'CONAN_'. Finally, this generator creates a file that can be included with make. So 'include conanbuild.make' will work and you don't need to source environment prior to running the build.

TBH, it would be nice if they just brought back the makefile generator support. If people aren't against it, maybe a variation of what I have done here could be added to conan. 

```
import re
from conan import ConanFile
from conan.tools.files import copy, save
from conan.tools.gnu import AutotoolsToolchain, AutotoolsDeps


class MakefileDeps:
    # conan 2.0 deprecated the makefile generator... so implementing a simple makefile generator
    def __init__(self, conanfile):
        self._conanfile = conanfile

    def generate(self):
        autotools = AutotoolsDeps(self._conanfile)
        env_vars = autotools.environment.vars(self._conanfile)
        autotools.generate()

        content = []

        # Rename the global CPPFLAGS, LDFLAGS, etc., with a "CONAN_" prefix --  CONAN_CPPFLAGS, CONAN_LDFLAGS, ...
        for key, value in env_vars.items("${name}"):
            value = re.sub(r"\$(\w+)", r"$(\1)", value)
            value = value.replace('"', '\\"')
            content.append(re.sub(f"{key}", f"CONAN_{key}", f"{key} := {value}"))

        # add CONAN_<PKG_NAME>_PKG_FOLDER := <path/to/package/folder> variables
        for require, dep in self._conanfile.dependencies.items():
            package_name = re.sub("-", "_", dep.ref.name).upper()
            if dep.package_folder:
                content.append(f"CONAN_{package_name}_PKG_FOLDER := {dep.package_folder}")
        save(self._conanfile, f"{self._conanfile.generators_folder}/conanbuild.make", "\n".join(content))

```

**Comment by iskunk** [comment]
@gabeblack: While one can get a lot of what's needed from the top-level package directories, you'd still want access to the `cpp_info` variables for more complex dependencies. Especially packages that make use of the [components](https://docs.conan.io/2/reference/conanfile/methods/package_info.html#components) abstraction.

All the same, thanks for sharing your code. I'm still finding my way around the codebase, and this gave me some helpful clues. I hope to have something to show here before too long.

**Comment by MachineCod** [comment]
Hello , 
Is there any updates on the issue ? 
I also need the make generator for my project and I do not want to use any workaround solutions that might break.

**Comment by iskunk** [comment]
Hi everyone,

I've put together a Make generator for Conan 2. For those interested, it is available in the [feature/make-generator](https://github.com/iskunk/conan/tree/feature/make-generator) branch of my Conan fork. Please feel free to give `MakeDeps` a try, and post your feedback. I will submit a PR for it in a few days.

Note that this is not a drop-in replacement for the Conan 1 `make` generator. The output filename is different (`conandeps.mk` instead of `conanbuildinfo.mak`) and there are some differences in the variable values (notably, components are now broken out into their own variables instead of being aggregated). The general approach is the same, however, so it should not take much work to adapt existing projects.

As a bonus, I've also included a `ShDeps` Bourne-shell generator. The logic is identical to `MakeDeps`; only the output syntax differs, and thus I set up the classes so that only a few minor overrides are needed.

**Comment by memsharded** [comment]
Hi everyone,

This is getting enough feedback to increase its priority, it seems that pure Makefile generators as built-in could make sense.
I am going to push it into our Conan 2 backlog. Thanks everyone for the contributions and code, we will use that as a starting point.

**Comment by iskunk** [comment]
Thank you @memsharded, it's good to see this issue getting traction.

Please give my implementation a try. I took the opportunity to not just recreate the old `make` generator, but improve upon it in several ways:

* Include settings in the output, so consumer makefiles can conditionalize on `$(CONAN_OS)` and the like;

* Avoid use of non-`[0-9A-Za-z_]` characters in variable names, like hyphens;

* Nicer formatting, e.g. properly indented continuation lines;

* Eliding variables without a value, to cut down on noise;

* Listing out components separately instead of aggregating them;

* `CONAN_*_FLAG` variables to easily allow adding `-D`/`-I`/`-L` prefixes to the values for direct use in compiler invocations;

* Minor path postprocessing to absorb long cache directory prefixes into the appropriate `CONAN_ROOT_*` variable, further cutting down on noise;

* Informative comments to help in navigating the file.

The old `make` generator always felt like a rush job. Here, I aimed to build something with better fit and finish overall.

**Comment by memsharded** [comment]
That is great, thanks for the explanation!

It seems that we can highly leverage it, maybe just the usual split between MakeDeps and MakeToolchain (the definition of settings for example, goes to the toolchain), but looks very good.

**Comment by reedhedges** [comment]
Tried @iskunk 's generator, works ok so far with conan 2.0.6 just by copying `makedeps.py` to `~/.conan2/extensions/generators`.  (Not sure how to use shdeps.py since it is importing makedeps.py, what is the right module path?)   I did rename it to MakefileDeps to have a bit less generic name.   

It would be nice to have toolchain flags generated as well.

Default (or aggregate?) `...LIBS...`, `...SYSTEM_LIBS...` and other such flags for all components of the dependency might also be good, not just the include and lib dirs.  (e.g. `CONAN_LIBS_FMT`, `CONAN_SYSTEM_LIBS_FMT` vs. using `CONAN_LIBS_FMT__FMT`),  I think the Makefile generator in conan1 did something like this?
The main/default library components seem to use identifiers with leading underscores, which results in awkward variable names like `CONAN_LIBS_FMT__FMT`.   What did the previous Makefile generator do? Was it just emitting a default variable name without component name from that default component?

One other problem that is now fixed is that I first tried it with an older version of conan2 before upgrading (sorry, forgot what the previous version was) and encountered a problem with `None` passed to `_makeify()`, perhaps because the catch2 package hadn't been fully transitioned to conan2? Or something in conan was fixed? Had warnings like :
```
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.filenames' used in: catch2/3.3.2
WARN: deprecated:     'cpp_info.names' used in: catch2/3.3.2, fmt/10.0.0
WARN: deprecated:     'cpp_info.build_modules' used in: catch2/3.3.2
```
I hacked makedeps.py to just skip components with no name but whatever the problem resulting in null names seems to have been fixed now. 

Anyway, thanks @iskunk !

**Comment by memsharded** [comment]
> One other problem that is now fixed is that I first tried it with an older version of conan2 before upgrading (sorry, forgot what the previous version was) and encountered a problem with None passed to _makeify(), perhaps because the catch2 package hadn't been fully transitioned to conan2? Or something in conan was fixed? Had warnings like :

The warnings are expected, the recipes in ConanCenter still keep some compatibility with Conan 1.X legacy generators, until that compatibility is removed, this warning will be there. The removal of support in ConanCenter for these will be gradual, it is important to give time to users to upgrade.

**Comment by iskunk** [comment]
> It seems that we can highly leverage it, maybe just the usual split between MakeDeps and MakeToolchain (the definition of settings for example, goes to the toolchain), but looks very good.

Ah, thanks. I was focused on the utility of having Make-accessible settings variables, but was unclear on the boundaries of where they should go.

> Default (or aggregate?) `...LIBS...`, `...SYSTEM_LIBS...` and other such flags for all components of the dependency might also be good, not just the include and lib dirs. (e.g. `CONAN_LIBS_FMT`, `CONAN_SYSTEM_LIBS_FMT` vs. using `CONAN_LIBS_FMT__FMT`), I think the Makefile generator in conan1 did something like this?

`CONAN_LIBS_FMT` and `CONAN_SYSTEM_LIBS_FMT` are unset because the package itself does not give them a value. What you're seeing with `CONAN_LIBS_FMT__FMT` having a value rather than `CONAN_LIBS_FMT` is similarly just the way the package is structured (which is arguably not ideal). Compare that to, say, `zlib/1.2.13`, which has `CONAN_LIBS_ZLIB` set, and no components. (IMO, for a simple package with a single library, the `zlib` approach is the right one.)

Note that the Conan 1 `make` generator didn't do components---it just aggregated everything for a given package together. This was unhelpful if your application needs to link against only a single component. (A good example of this is `pcre2/10.42`, which offers 8-, 16-, and 32-bit versions of the library, each residing in its own Conan component. Most applications use only one of these.)

> The main/default library components seem to use identifiers with leading underscores, which results in awkward variable names like `CONAN_LIBS_FMT__FMT`. What did the previous Makefile generator do? Was it just emitting a default variable name without component name from that default component?

I thought the normal practice was for the "default" component to have the same name as the package? E.g. I would expect `fmt::fmt`. Is there some kind of convention behind the use of `fmt::_fmt` instead? That seems incorrect (or at least questionable) to me, from what I've seen.

In any event, not only did the old `make` generator aggregate components together, it didn't bother to sanitize the variable names. So if you had a dependency `foo-bar`, you would get variables like `CONAN_LIBS_FOO-BAR` :stuck_out_tongue:  (GNU Make can handle that, but not regular Make.) Hence why it always felt like a rush job...

**Comment by reedhedges** [comment]
I agree that being able to use flags for each component on its own is important.  (This is also analogous to how you would use it in CMake, e.g. `fmt::fmt`)  As is sanitizing the names better.

Would it be appropriate for the generator to print output indicating how the variables are named?  (Not that it's hard to just look at the generated conandeps.mk file, just would remove one small potential stumbling block for someone.) 

Don't know much about the details of Conan packages, I only ever use them in some projects in some basic ways.

**Comment by iskunk** [comment]
> Would it be appropriate for the generator to print output indicating how the variables are named? (Not that it's hard to just look at the generated conandeps.mk file, just would remove one small potential stumbling block for someone.)

That would end up spamming the user for something like `qt/5.15.9` :-]

I think it's enough just to review the file. A lot of the changes I made were with an eye to making the output easier to read/navigate.

**Comment by iskunk** [comment]
To everyone watching this issue: In case you missed it, PR #14133 with a new Make dependency generator is awaiting final review. Be sure to give it a try, and voice any concerns you may have, before it gets merged.

**Comment by tbsuht** [comment]
> Tried @iskunk 's generator, works ok so far with conan 2.0.6 just by copying `makedeps.py` to `~/.conan2/extensions/generators`. (Not sure how to use shdeps.py since it is importing makedeps.py, what is the right module path?) I did rename it to MakefileDeps to have a bit less generic name.

This is not working for me, did something change to test this?
```
$ ls ~/.conan2/extensions/generators/           
makedeps.py
```

Running `conan create .` on a conanfile which defines `generators = "MakeDeps"`:
`ERROR: Invalid generator 'MakeDeps'. Available types: CMakeToolchain, CMakeDeps, MesonToolchain, MSBuildDeps, MSBuildToolchain, NMakeToolchain, NMakeDeps, VCVars, QbsProfile, VirtualRunEnv, VirtualBuildEnv, AutotoolsDeps, AutotoolsToolchain, PkgConfigDeps, BazelDeps, BazelToolchain, IntelCC, XcodeDeps, XcodeToolchain, PremakeDeps`

**Comment by iskunk** [comment]
Hi @tbsuht,

I never understood having my implementation run out of the `generators` folder, and it's old news at this point anyway. Go to @uilianries's tree and feature branch [here](https://github.com/uilianries/conan/tree/feature/make-generator), and give that a try. That's the latest-and-greatest code that is being brushed up for merging into Conan.

**Comment by tbsuht** [comment]
@iskunk Thanks, that helped!

Is there a small example available with the new generator? I tried to get my [own example](https://github.com/tbsuht/conan-makefiles) running by just exchanging the AutoTools stuff with the new generator but didn't get it running.

/e: maybe that has nothing to do with the new generator...?

**Comment by uilianries** [comment]
> @iskunk Thanks, that helped!
> 
> Is there a small example available with the new generator? I tried to get my [own example](https://github.com/tbsuht/conan-makefiles) running by just exchanging the AutoTools stuff with the new generator but didn't get it running.
> 
> /e: maybe that has nothing to do with the new generator...?

@tbsuht The MakeDeps is under development right now. I recommend you following the PR #14133 and take a look in the tests.

**Comment by iskunk** [comment]
> Is there a small example available with the new generator? I tried to get my [own example](https://github.com/tbsuht/conan-makefiles) running by just exchanging the AutoTools stuff with the new generator but didn't get it running.

Are you running into an error of some kind?

FWIW, once you have that feature branch up and running, the following command (invoked in an empty directory) will have the new generator produce some sample output:
```
conan install -g MakeDeps --requires=pcre2/10.42
```
(Your example does not appear to declare any dependencies, so I suspect the new generator isn't producing meaningful output for it.)

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/14133 provided ``MakeDeps``, will be released in next 2.0.10. Feedback will be very welcome. The ``MakeToolchain`` will be next, but lets stabilize ``MakeDeps`` a bit and learn from it before ``MakeToolchain``.
